File: python/src/bytifi/base64url.py
import base64
import re

_BASE64URL_CHARSET = re.compile(r"^[A-Za-z0-9_-]*$")


def to_base64url(data: bytes) -> str:
    encoded = base64.b64encode(data).decode("ascii")
    return encoded.replace("+", "-").replace("/", "_").rstrip("=")


def from_base64url(value: str) -> bytes:
    normalized = str(value or "")
    if not _BASE64URL_CHARSET.fullmatch(normalized):
        raise ValueError("Invalid base64url string: contains disallowed characters.")

    padded = normalized.replace("-", "+").replace("_", "/")
    padded += "=" * ((4 - len(padded) % 4) % 4)
    return base64.b64decode(padded)

File: lib/base64url.js
const BASE64URL_CHARSET = /^[A-Za-z0-9_-]*$/

export function toBase64Url(bytes) {
  const buffer = Buffer.isBuffer(bytes) ? bytes : Buffer.from(bytes)
  return buffer
    .toString('base64')
    .replace(/\+/g, '-')
    .replace(/\//g, '_')
    .replace(/=+$/g, '')
}

export function fromBase64Url(value) {
  const normalized = String(value || '')

  if (!BASE64URL_CHARSET.test(normalized)) {
    throw new Error('Invalid base64url string: contains disallowed characters.')
  }

  const padded = normalized.replace(/-/g, '+').replace(/_/g, '/')
    .padEnd(normalized.length + ((4 - (normalized.length % 4)) % 4), '=')
  return Buffer.from(padded, 'base64')
}
